infrastructure/async/src/main/java/tech/pegasys/teku/infrastructure/async/SafeFuture.java

```diff
@@ -192,6 +193,28 @@ public static <T> SafeFuture<List<T>> collectAll(final SafeFuture<T>... futures)
         .thenApply(__ -> Stream.of(futures).map(SafeFuture::join).collect(toList()));
   }
 
+  /**
+   * Waits for all the supplied futures to complete then returns a single future that combines all
+   * the successful results.
+   *
+   * <p>If all the futures complete successfully the returned future completes with a {@link List}
+   * of each result in the same order as the futures were given.
+   *
+   * <p>If any futures complete exceptionally the returned future completes with a {@link List} of
+   * each result in the same order as the futures were given filtering out the exceptionally
+   * completed futures.
+   *
+   * @param futures the futures to collect results from
+   * @param <T> the result type to collect
+   * @return a new future that completes when all the supplied futures complete
+   */
+  public static <T> SafeFuture<List<T>> collectAllSuccessful(final Stream<SafeFuture<T>> futures) {
+    final Stream<SafeFuture<Optional<T>>> optionalFutures =
+        futures.map(future -> future.thenApply(Optional::of).exceptionally(__ -> Optional.empty()));
```

Comment:
what about having an optional callback to be able to log errors? Or you can return a Pair of future lists, successful and failed list

Reply:
but you're intercepting only when the signer fails. If the exception occurs in some other places the error is lost